### ISSUE TITLE
Feature Idea: Remove comments

### ISSUE THREAD

**Issue by hellp** [coralproject/talk#1534]
# Feature Idea: Remove comments

## Use case

- As a normal user (commenter) I want to remove my own comments.
- As a moderator I want to remove user comments while still keeping them in the stream as an "anchor" for replies etc.

While this looks like 2 use-cases, from our point of view it is one. We want a feature where *removing* an *accepted* comment keeps it visible in the stream while overwriting its content with a remark of why it was removed ("tombstoning"), e.g. "The user deleted this comment." or "This comment was removed due to legal reasons." etc.

I found that this idea has already been discussed as part of the following bigger story: [#154026776 DELETE ME | As a commenter, I want to delete my profile and all associated comments](https://www.pivotaltracker.com/story/show/154026776). There the following questions regarding the feature came up:

> MOD SIDE:
>
> - How do we handle deleted comments - mod side
> - …
>
> COMMENTER-SIDE:
>
> - How do we handle deleted comments - commenter side
> - Tombstoning?? "This comment has been deleted" - continue to show
>   replies/conversation/etc - show replies for permalinks too

So before we start implementing it on our own, we'd like to get some feedback here. It seems that it cannot/should not be done as a plugin, as it likely touches too many core concepts and mechanisms in core. So, from our POV, it would be nice to get the feature into core.

Here are some of our open questions:

- Would `REMOVED` be a comment status just as `ACCEPTED` and `REJECTED`? Or should it be a flag orthogonal to these?
  (I feel that implementing it as a status would be cleaner, but also more invasive.)

- Is the behaviour clear from the user perspective? For example: what is the difference between a *removed* comment and a comment that was first *accepted* and later *rejected*? Why is the *removed* comment still in the stream (as a 'tombstone stub' and anchor for replies), while the rejected comment is completely gone? — We figured that both should be possible at the moderator's discretion. Comments that obviously (whatever that means in each community) violate the rules will still be rejected; comments that are removed for not-so-obvious reasons and/or that already spawned a lively reply-thread will only be tombstoned.

- For the tombstone texts we would probably provide a set of templates that a moderator can choose (legal reason, generic policy violation). Should these texts be translatable? Or is it fine to have them configured somewhere in the site's main language?

tldr;

- How likely is it that this feature will make it into core?
- Can you give us implementation hints how to do it best?

**Comment by kgardnr** [comment]
Thank you for this detailed feature suggestion @hellp! I will review with the team and get back to you.

**Comment by kgardnr** [comment]
Hey @hellp! Got feedback from our team - we think this is a great idea to introduce into core. Here is how we think we would approach this - broken into 4 stories - 2 based on above and 2 additional enhancements:

**DELETE MY COMMENT** | As a commenter, I want to delete my comment so that it doesn't show publicly anymore

    - Creates a tombstone
    - The Delete button is probably in the Edit button area
    - Single default tombstone language
    - Moderator can see the text of the comment that was removed in the user drawer
    - Replies stay visible


**MOD TOMBSTONE OPTION** | As a moderator, I want the ability to tombstone comments I reject so that I can be transparent about my actions. 

    - Dropdown template language reject options instead of the single Reject button
    - Option chosen is shown in the comment card in the admin panel
    - Replies stay visible

**HIDE DELETED** | As a readeer, I want to hide all deleted comments and their replies so that my reading experience isn’t too cluttered by things I can’t read

    - Add ‘Hide deleted comments and their replies’ to Viewing Options, like Off Topic

**CONFIG DELETE** | As an org, I want comment deletion to be configurable, so I can decide how this functions

    - Add to Configure page "Allow commenters to delete their comments" (site-wide)
    - Add to Stream configure "Allow commenters to delete their comments" (stream-only)
    - Our default for these would be 'off'

**Comment by hellp** [comment]
The tombstone templates would be configured/stored in `<root>/config.js` for now, right? I.e. configurable per instance, but static.

**Comment by hellp** [comment]
When keeping the original comment available for moderators to read, where would we store both texts? 

We thought of this approach: copy `comment.body` to `comment.metadata.removed_body` and then replace `comment.body` with the tombstone template. We feel that it's more "secure" from a privacy POV to get the offensive text out of `comment.body` forever.

- Or would we rather add a distinct new field à la `comment.removed_body`? 
- Or add a new property `tombstoned_by=<reason>` and let the UI decide what to show while keeping `comment.body` intact with the original text?

**Comment by wyattjoh** [comment]
We've implemented a piece of this in our GDPR feature for deleting the users whole accounts: https://github.com/coralproject/talk/pull/1536 but we don't preserve any of the data in a `removed...` field, instead we purge all the associated data permanently. This to us represents the safest from a privacy standpoint.

I also feel that the rejecting feature best suits the needs for moderators to remove a comment from the comment stream instead of giving them the ability to remove the comment because again, when deleting a comment, we purge all the data permanently.

**Comment by hellp** [comment]
> I also feel that the rejecting feature best suits the needs for moderators to remove a comment from the comment stream instead of giving them the ability to remove the comment because again, when deleting a comment, we purge all the data permanently.

@wyattjoh 
Sorry, I don't get this part.

You think that Reject should remain the only possibility for moderators to remove user comments? What about the remove-with-a-note feature? (Maybe it's just my confusion with all the remove/delete/reject terminology.)

**Comment by wyattjoh** [comment]
In Talk, there are two ways that prevent comments from being visible to normal users:

1. **Rejecting** a comment
2. **Deleting** a comment

Rejecting a comment rather than deleting it has some advantages:

- We retain the user details on who wrote the comment
- We retain the comment text
- We hide all replies to the rejected comment, which often times will devolve and lose context as a conversation progresses

Currently, the only planned way to delete a comment is through the ability for users to delete their entire account, which will "tombstone" their existing comments. It's not due to a moderation reason that they deleted their account, instead it's a privacy/data control feature. For that reason, replies would still be visible.

For a those reasons, I think moderators would be better off rejecting a comment rather than deleting a comment. We have a host of tools developed for managing a users reputation that would allow moderators to make better decisions in the future based on the reject/approve flow that would be lost if a comment was deleted.

This is unrelated I think from the users ability to delete their own comments, which is a feature that I think we can explore in the future for sure.

**Comment by tcmal** [comment]
So if the comment is deleted because someone deleted their accuont, that should probably be a different state than a user deleted comment as the original body isn't available at all anymore.

**Comment by wyattjoh** [comment]
Exactly @tcmal, we're in the process of exploring that, as there's still lots of questions to answer for if a user deletes their own comment:

- Do we keep the reference to the user to show that it was them who authored it?
- Do we show replies to the deleted comment?

It really depends on the use case, but we can update this ticket as we explore the feature.

**Comment by hjanuschka** [comment]
we implemented a custom endpoint, in case the user requests to delete himself, #gdpr.
created a custom user "removed-user" -  and comments get migrated to that virtual user.

**Comment by AbodyAmer** [comment]
hi @kgardnr 
is deleting comment feature already has been implemented?? 
please tell me if you have any plug-in that allow me to delete a comment. 
Thank you

**Comment by kgardnr** [comment]
Hey @AbodyAmer no we don't support this currently - this issue is that open feature request.

Moderators can reject comments and users can delete their accounts, but users cannot remove a single comment. See above discussion.

**Comment by JulianaBesser** [comment]
Is there an update to the Feature?

The problem is that we have comments from a deleted user in den pending view and we can't reject them. 
We just want to get them out of the pending view.

![Bildschirmfoto 2022-01-25 um 13 17 27](https://user-images.githubusercontent.com/8426266/150975658-95034770-7c51-44cc-88ae-d908a627488b.png)
